File: Graph.cpp
//------------------------------------------------------------------------------
//  Graph.cpp
//  Created by Kunlakan Cherdchusilp, Ngoc Luu, Jonathan Earl, and Alan Nyugen
//------------------------------------------------------------------------------
// Graph is an undirected and unweigthed graph that is represented by adjacency
// list of vertices that are connected to each other.
// features are included:
//   -- allows adding and removing edges
//   -- allows displaying of the whole Graph with distance and path
//
// ASSUMPTIONS:
//   -- No graph ever share the same object (they could share the same value)
//   -- file input must be properly formatted:
//          1   2
//          2   4       The frist number is the vertex from
//          3   1       The second number is the vertex to
//
//------------------------------------------------------------------------------


#include "Graph.h"

//--------------------------- A Default Constructor ----------------------------
// Default constructor for class Graph
// Preconditions: None
// Postconditions: None
Graph::Graph(){}

//--------------------------------- Destructor ---------------------------------
// Destructor for class Graph
// Preconditions: None
// Postconditions: None
Graph::~Graph()
{}


//--------------------------------- buildGraph ---------------------------------
// Builds a graph by reading data from an ifstream
// Preconditions:  infile has been successfully opened and the file contains
//                 properly formated data (according to the program specs)
// Postconditions: A graph is read from infile and stored in the object
void Graph::buildGraph(ifstream& infile)
{
    int src = -1, dest = -1;
    
    infile >> src >> dest;
    
    while (infile)
    {
        if(src != dest)
        {
            exist(src);
            exist(dest);
            
            vertices[src].insert(dest);
            vertices[dest].insert(src);
        }
        
        infile >> src >> dest;
    }
}

//------------------------------- PRIVATE: exist -------------------------------
// Check if vertex already exists in the vector vertices
// Preconditions: None
// Postconditions: If vertex already exists in the vector vertices, then do
//                 nothing. Otherwise, add vertex to the vector vertices.
void Graph::exist(const int &vertex){
    if(vertex < vertices.size())
        return;
    
    for(long i = vertices.size(); i <= vertex; i++)
        vertices.push_back(*new unordered_set<int>);
    
    vertices[vertex] = *new unordered_set<int>;
}


//---------------------------------- display -----------------------------------
// Display a all detailed path
// Preconditions: vertices[vertexFrom] and its data must exist
// Postconditions: all detailed path is displayed
void Graph::display() const
{
    cout << "From\t\t";
    cout << "To" << endl;
    
    for(int i = 0; i < vertices.size(); i++)
    {
        if (vertices[i].size() > 0)
        {
            cout << i << ": ";
            
            for (int vertice : vertices[i])
                cout << vertice << " ";
            
            cout << endl;
        }
    }
}

//------------------------------ enumerateSubgraph -----------------------------
// Enumerate size-k subgraphs of the original graph
// Preconditions: The graph should have already been built or exists
// Postcondition: The list of subgraphs are displayed
void Graph::enumerateSubgraph(const int &k)
{
    count = 0;
    
    for(int i = 0; i <= vertices.size(); i++)
    {
        if(vertices[i].size() > 0)
        {
            list<int> Vsubgraph;
            Vsubgraph.push_back(i);
            
            unordered_set<int> visited;
            visited.insert(i);
            
            unordered_set<int> Vextension;
            getExtension(Vextension, i);
            
            extendSubgraph(Vsubgraph, Vextension, visited, i, k);
        }
    }
    cerr << count << endl;
}

//--------------------------- PRIVATE: getExtension ----------------------------
// Add all neighbors of vertex to Vextension
// Precondition: Vextension must be decalired and intialized before calling
//               this function.
// Postcondition: All neighbors of vertex are added to Vextension
void Graph::getExtension(unordered_set<int> &Vextension, const int &vertex)
{
    for (int i : vertices[vertex])
    {
        if (i > vertex)
            Vextension.insert(i);
    }
}


//--------------------------- PRIVATE: extendSubgraph --------------------------
// Recursively looking size-k subgraphs of the graph.
// Precondition: The graph should have already been built or exists
// Postcondition: The list of subgraphs are displayed
void Graph::extendSubgraph(list<int> &Vsubgraph, unordered_set<int> &Vextension, unordered_set<int> &visited, const int &v, const int &k)
{
    if(Vsubgraph.size() == k-1)
    {
        for(int w : Vextension)
            count++;
        
        return;
    }
    
    list<int> unvisit;
    
    while(Vextension.size() != 0)
    {
        int w = *Vextension.cbegin();
        Vextension.erase(w);
        
        Vsubgraph.push_back(w);
        visited.insert(w);
        
        unvisit.push_back(w);
        
        unordered_set<int> Vextension2 = unordered_set<int>(Vextension);
        
        for (int vertex : vertices[w])
        {
            if (vertex > v && visited.count(vertex) == 0 && Vextension.count(vertex) == 0)
                Vextension2.insert(vertex);
        }
        
        extendSubgraph(Vsubgraph, Vextension2, visited, v, k);
        Vsubgraph.pop_back();
    }

    for(int i : unvisit)
        visited.erase(i);

}

File: NemoSQL_Python/Graph.py
import sqlite3
import datetime

connection = sqlite3.connect(':memory:')

k = 0
count = 0
table_names = {}
cursor = connection.cursor()
visited = set()
vertices = []

# ------------------------------------ main ------------------------------------
def main():
    buildGraph("./big_ecoli.txt")

    # start = datetime.datetime.now()
    # enumerateSubgraph(3)
    # end = datetime.datetime.now()
    # print "Time : ", (end - start).total_seconds() * 1000

    start = datetime.datetime.now()
    enumerateSubgraph(4)
    end = datetime.datetime.now()
    print "Time : ", (end - start).total_seconds() * 1000

    # start = datetime.datetime.now()
    # enumerateSubgraph(5)
    # end = datetime.datetime.now()
    # print "Time : ", (end - start).total_seconds() * 1000


# --------------------------------- buildGraph ---------------------------------
# Builds a graph by reading data from the given file
# Preconditions:    file has been successfully opened and contains property
#                   formatted data (according to the program specs)
# Postconditions:   A graph is read from infile and stored in the object
def buildGraph(file):
    global vertices

    cursor.execute("CREATE TABLE VERTICES (node INT NOT NULL, neighborTable TEXT NOT NULL, PRIMARY KEY(node))")

    with open(file, "r") as inputFile:
        for line in inputFile:
            data = line.strip().split()
            nodeFrom = data[0]
            nodeTo = data[1]

            if nodeFrom not in table_names:
                table_names[int(nodeFrom)] = "SELECT neighbor FROM _" + nodeFrom + " WHERE neighbor > ?"
            if nodeTo not in table_names:
                table_names[int(nodeTo)] = "SELECT neighbor FROM _" + nodeTo + " WHERE neighbor > ?"

            neighborTableName = nodeFrom
            cursor.execute("INSERT OR IGNORE INTO VERTICES VALUES (?, ?)", (nodeFrom, neighborTableName))
            cursor.execute("CREATE TABLE IF NOT EXISTS _" + neighborTableName + " (neighbor INT NOT NULL, PRIMARY KEY(neighbor))")
            cursor.execute("INSERT OR IGNORE INTO _" + neighborTableName + " VALUES (?)", [nodeTo])

            neighborTableName = nodeTo
            cursor.execute("INSERT OR IGNORE INTO VERTICES VALUES (?, ?)", (nodeTo, neighborTableName))
            cursor.execute("CREATE TABLE IF NOT EXISTS _" + neighborTableName + " (neighbor INT NOT NULL, PRIMARY KEY(neighbor))")
            cursor.execute("INSERT OR IGNORE INTO _" + neighborTableName + " VALUES (?)", [nodeFrom])

    cursor.execute("SELECT node FROM VERTICES")
    vertices = cursor.fetchall()





# ------------------------------ enumerateSubgraph -----------------------------
# Enumerate size-k subgraphs of the original graph
# Preconditions: The graph should have already been built or exists
# Postcondition: The number of subgraphs found is displayed
def enumerateSubgraph(sizeK):
    global count, k, visited, vertices

    k = sizeK
    count = 0
    visited = set()

    # test hash table vs sql query table
    # ************** hash table ***************
    # for v in table_names:
    #     Vextension = getExtension(v)
    #     visited.add(v)
    #     extendSubgraph([v], Vextension)
    #     visited.remove(v)

    # *********** sql query table **************
    for v in vertices:
        Vextension = getExtension(v[0])
        visited.add(v[0])
        extendSubgraph([v[0]], Vextension)
        visited.remove(v[0])

    print count


# -------------------------------- getExtension --------------------------------
def getExtension(w, v = None, Vextension = set()):
    global visited, table_names

    newExtension = set(Vextension)

    if v is None:
        v = w
    # Test construct string vs look up string
    # ***********   using hash_table   ****************
    cursor.execute(table_names[w], [v]) # marginally faster ~2.5% - 5%
    # ***********   constructing string ***************
    # cursor.execute("SELECT neighbor FROM _" + str(w) + " WHERE neighbor > ?", [v])

    allNeighbor = cursor.fetchall()

    for neighbor in allNeighbor:
        if neighbor[0] not in visited:
            newExtension.add(neighbor[0])

    return newExtension


# ------------------------------- extendSubgraph -------------------------------
# Recursively looking size-k subgraphs of the graph.
# Precondition: The graph should have already been built or exists
# Postcondition: The number of subgraphs found is displayed
def extendSubgraph(Vsubgraph, Vextension):
    global count, visited

    if len(Vsubgraph) == k-1 :
        for w in Vextension:
            count += 1
        return

    unvisited = []

    while len(Vextension) > 0:
        w = Vextension.pop()

        Vsubgraph.append(w)
        unvisited.append(w)
        visited.add(w)

        Vextension2 = getExtension(w, Vsubgraph[0], Vextension)

        extendSubgraph(Vsubgraph, Vextension2)
        Vsubgraph.pop()

    for i in unvisited:
        visited.remove(i)



main()